src/PIL/Image.py

```diff
@@ -528,14 +528,15 @@ def __enter__(self):
         return self
 
     def __exit__(self, *args):
-        if hasattr(self, "fp") and getattr(self, "_exclusive_fp", False):
-            if getattr(self, "_fp", False):
-                if self._fp != self.fp:
-                    self._fp.close()
-                self._fp = DeferredError(ValueError("Operation on closed image"))
-            if self.fp:
-                self.fp.close()
-        self.fp = None
+        if hasattr(self, "fp"):
+            if getattr(self, "_exclusive_fp", False):
+                if getattr(self, "_fp", False):
+                    if self._fp != self.fp:
+                        self._fp.close()
+                    self._fp = DeferredError(ValueError("Operation on closed image"))
+                if self.fp:
+                    self.fp.close()
+            self.fp = None
```

Comment:
Might be easier to read if the outermost layer of the pyramid was negated?
```suggestion
        if not hasattr(self, "fp"):
            return
        if getattr(self, "_exclusive_fp", False):
            if getattr(self, "_fp", False):
                if self._fp != self.fp:
                    self._fp.close()
                self._fp = DeferredError(ValueError("Operation on closed image"))
            if self.fp:
                self.fp.close()
        self.fp = None
```

Reply:
Easier to read in isolation, yes, but I kind of like the fact that this version mirrors `close()` - https://github.com/python-pillow/Pillow/blob/5431b15bd2f8177b762332ac632b0342c3b9d506/src/PIL/Image.py#L530-L565

Reply:
Should that code be extracted from `close()` then and reused in both of these places? `_close_fp()`?

Reply:
Ok, I've pushed a commit.